qiskit/circuit/instructionset.py

```diff
@@ -58,17 +162,34 @@ def inverse(self):
             self.instructions[index] = instruction.inverse()
         return self
 
-    def c_if(self, classical, val):
-        """Add condition on classical register to all instructions."""
-        if self.circuit_cregs:
-            circuit_clbits = [bit for creg in self.circuit_cregs for bit in creg]
-        if isinstance(classical, (Clbit, int)):
-            if isinstance(classical, int):
-                classical = circuit_clbits[classical]
-            for creg in self.circuit_cregs:
-                if creg[:] == [classical]:
-                    classical = creg
-                    break
+    def c_if(self, classical: Union[Clbit, ClassicalRegister, int], val: int) -> "InstructionSet":
+        """Set a classical equality condition on all the instructions in this set between the
+        :obj:`.ClassicalRegister` or :obj:`.Clbit` ``classical`` and value ``val``.
+
+        .. note::
+            This overrides any previously set condition.
```

Comment:
Maybe expand on this a bit and say it will happen silently too. I just don't want people to be surprised that if you do `qc.x(0).c_if(qr, 2).c_if(qr[0], 1)` that there is only a condition on the first bit.

Reply:
Sure - this isn't changed behaviour, it's just adding a documentation note on something that was already happening silently, so noisy is better.  With the current way `Instruction.condition` works, we can't really support compound conditions at all, but when we get a bit more QASM3-ish support, we'll need to overhaul that field/remove it in favour of `IfElseOp` and we can have a proper boolean expression in some form.

Done in e0b3700ae.